### DISCUSSION TITLE
Strategy for working with components that don't expose an input

### DISCUSSION THREAD

**Discussion by erik-stephens** [react-hook-form/react-hook-form#4566] [Q&A]
Apologies if this is a FAQ.  I don't know how to inform RHF of the input ref to `focus()` on when there are validation errors in components like react-date-picker and react-bootstrap's ToggleButton.  I imagine this is a common issue with complicated components.  Is there a general strategy I can apply without having to get the component to alter its api?  Thank you!

**Comment by bluebill1049** [comment]
use `Controller` or `useController`. There is onFocus prop for Controller as well.

**Reply by erik-stephens** [reply]
I've been using Controller.  It's helped fill a lot of the gaps, so many thanks for that!  I worry that I may be completely misunderstanding refs & focusing because I can't find much discussion.  I can't get these components to get the focus when their validations fail:

- [react-bootstrap ToggleButton](https://react-bootstrap.github.io/components/buttons/#toggle-button-props) - it provides an `inputRef` prop but doesn't work as I expected
- [react-date-picker](https://github.com/wojtekmaj/react-date-picker#user-guide) - it doesn't provide any property that I can find

Is the following statement correct?

"If the component does not provide a way to give RHF a reference to the underlying input element, then the issue is with the component and not RHF."

Any workarounds for such components?  I'm trying to figure out where my focus (no pun intended) should be.  Thank you!

**Reply by bluebill1049** [reply]
you will need to inspect what's the inputRef that they r exposing. some of them are an object which may contain the input ref, it's not guarantee that the component do expose input's ref.

**Reply by xddz9** [reply]
> use `Controller` or `useController`. There is onFocus prop for Controller as well.

@bluebill1049 Now that v7 removed onFocus from Controller, is there another way to handle this? I know we have a ref props now, but it seems to work for inputs only. What if, for instance, I want to have focus instead on a button in case of a validation error?

Thanks! :)

**Reply by bluebill1049** [reply]
`ref` is basically the `onFocus` now. 

```
<Controller render={( field: { ref }) => {}} /> // ref.focus will be invoked when an error occurred.
```
